src/v/storage/disk_log_impl.cc

```diff
@@ -1731,6 +1733,446 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
       .then([this](auto rdr) { return _readers_cache->put(std::move(rdr)); });
 }
 
+namespace details {
```

Comment:
Thank you for pulling this logic into the storage layer! Now there are a couple of clearly separate pieces to this PR: local storage changes, and archival changes. Would you mind splitting them into separate PRs, where the latter depends on the former? That will make them both more reviewer-friendly

Reply:
done

Reply:
Thank you!